src/validator/periods/submitPeriod.js

```diff
@@ -0,0 +1,110 @@
+ */
+
+const {
+  getSlotsByAddr,
+  sendTransaction,
+  buildCas,
+} = require('../../utils');
+const { logPeriod } = require('../../utils/debug');
+const checkEnoughVotes = require('./checkEnoughVotes');
+const submitPeriodVote = require('./submitPeriodVote');
+const isAlreadyVoted = require('./isAlreadyVoted');
+
+module.exports = async (
+  periodProposal,
+  bridgeState,
+  opts = {}
+) => {
+  const defaultResponse = { receiptPromise: Promise.resolve() };
+  const { proposerSlotId, blocksRoot, prevPeriodRoot } = periodProposal;
+  const { currentState, lastBlocksRoot, lastPeriodRoot } = bridgeState;
+  const { slots } = currentState;
+
+  logPeriod(
+    '[submitPeriod] proposerSlot:%d blocksRoot:%s prevPeriodRoot:%s',
+    proposerSlotId, blocksRoot, prevPeriodRoot
+  );
+
+  if (lastBlocksRoot === blocksRoot) {
+    // check if the period is already onchain
+    const submittedPeriod = await bridgeState.bridgeContract.methods
+        .periods(lastPeriodRoot)
+        .call();
+    
+    if (submittedPeriod.timestamp === '0') {
+      throw new Error('No period found onchain for bridgeState.lastBlocksRoot');
+    }
+
+    logPeriod('[submitPeriod] already seen onchain', lastPeriodRoot, submittedPeriod);
+    return { receiptPromise: Promise.resolve({ status: true }) };
+  }
+
+  const mySlots = getSlotsByAddr(slots, bridgeState.account.address);
+
+  if (!mySlots.length) {
+    logPeriod('[submitPeriod] No slots. Not a validator');
+    return defaultResponse;
+  }
+
+  const mySlotToSubmit = mySlots.find(slot => slot.id === proposerSlotId);
+
+  if (!isAlreadyVoted(blocksRoot, mySlots[0].id, periodProposal)) {
+    await submitPeriodVote(blocksRoot, periodProposal, bridgeState);
+  }
+  
+  // check if it is our turn to submit period
+  if (!mySlotToSubmit) {
+    logPeriod('[submitPeriod] Not a proposer. Skipping');
+    return defaultResponse;
+  }
+
+  // check if we have enough period votes for submission
+  const { result, votes, needed } = checkEnoughVotes(
+    blocksRoot, periodProposal, bridgeState.currentState.slots
+  );
+  
+  if (!result) {
+    logPeriod(
+      `[submitPeriod] Not enough period votes collected: ${votes}/${needed}. Waiting..`
+    );
+    return defaultResponse;
+  }
+
+  const cas = buildCas(periodProposal.votes);
+  logPeriod('[submitPeriod] CAS:%s',`0x${cas.toString(16)}`);
+
+  if (periodProposal.txHash) {
+    logPeriod('[submitPeriod] Already submitted. txHash: %s', periodProposal.txHash);
+    return defaultResponse;
+  }
+
+  const { receiptPromise } = await sendTransaction(
+    bridgeState.web3,
+    bridgeState.operatorContract.methods.submitPeriodWithCas(
+      mySlotToSubmit.id,
+      prevPeriodRoot,
+      blocksRoot,
+      `0x${cas.toString(16)}`
+    ),
+    bridgeState.operatorContract.options.address,
+    bridgeState.account,
+    opts
+  );
+
+  receiptPromise.on('transactionHash', (hash) => {
+    logPeriod('[submitPeriod] txHash: ', hash);
+    periodProposal.txHash = hash;
```

Comment:
recording the hash so the node knows it already submitted the period. Note that this is not a shared state (other nodes won't know about it), it is just for the proposer node not to propose multiple times while the submission is in flight

Reply:
See https://github.com/leapdao/leap-node/pull/379/files#diff-5e72911c8438e6f450b0d8caf9f3087fR81 and https://github.com/leapdao/leap-node/pull/379/files#diff-64e8a844a27bd8005d649e916b073b3eR18